Refactor param binding failure handling in RequestDelegate

marcominerva opened the pull request:
## Description

Refactor `RequestDelegateFactory` to handle parameter binding failures more flexibly:
- Execute filter pipelines on failure if filters are present and return type is `ValueTask<object?>`, allowing filters to observe/modify the 400 response.
- Short-circuit with a 400 response if no filters are present.
- Simplify logic using `Expression.Condition` and `Expression.Block`.

Add new test cases in `RouteHandlerEndpointRouteBuilderExtensionsTest.cs`:
- Validate behavior with/without filters and custom responses.
- Ensure handlers are not executed on binding failure.

Fixes #64341

marcominerva committed:
Refactor param binding failure handling in RequestDelegate

Refactor `RequestDelegateFactory` to handle parameter binding
failures more flexibly:
- Execute filter pipelines on failure if filters are present
  and return type is `ValueTask<object?>`, allowing filters
  to observe/modify the 400 response.
- Short-circuit with a 400 response if no filters are present.
- Simplify logic using `Expression.Condition` and
  `Expression.Block`.

Add new test cases in `RouteHandlerEndpointRouteBuilderExtensionsTest.cs`:
- Validate behavior with/without filters and custom responses.
- Ensure handlers are not executed on binding failure.

captainsafia commented:
@marcominerva To sanity check this, can you add tests for the short-circuiting logic in the infrastructure that is shared between RDG/RDF? That'll allow us to confirm that the behavior is the same across runtime and compile-time generated handler implementations.

The shared tests are [implemented here](https://github.com/dotnet/aspnetcore/tree/main/src/Http/Http.Extensions/test/RequestDelegateGenerator).

marcominerva committed:
Add tests for the short-circuiting logic in the infrastructure shared between RDG/RDF

This confirm that the behavior is the same across runtime and compile-time generated handler implementations.

marcominerva commented:
@captainsafia, I have added the new tests.

BrennanConroy commented on src/Http/Http.Extensions/src/RequestDelegateFactory.cs:
```diff
@@ -993,41 +993,31 @@ private static Expression CreateParamCheckingResponseWritingMethodCall(Type retu
+        Expression failureBranch;
+        if (factoryContext.EndpointBuilder.FilterFactories.Count > 0 && returnType == typeof(ValueTask<object?>))
+        {
+            failureBranch = Expression.Block(
```
nit: Could we keep (and update) the comments that show what the expression tree is generating? Very helpful for visualizing what's happening 😄

marcominerva committed:
Clarify param check failure logic with comments

Added explanatory comments to RequestDelegateFactory describing the checkWasParamCheckFailure conditional expression. These comments clarify how parameter binding or validation failures result in a 400 response, while successful checks invoke the handler method. No functional changes were made.

marcominerva committed:
Fix parameter binding failure handling with endpoint filters to ensure proper response status
